src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LspEditorFeatureDetector.cs

```diff
@@ -0,0 +1,102 @@
+
+[Export(typeof(ILspEditorFeatureDetector))]
+internal sealed class LspEditorFeatureDetector : ILspEditorFeatureDetector, IDisposable
+{
+    private readonly IUIContextService _uiContextService;
+    private readonly JoinableTaskFactory _jtf;
+    private readonly CancellationTokenSource _disposeTokenSource;
+    private readonly AsyncLazy<bool> _lazyUseLegacyEditorTask;
+
+    [ImportingConstructor]
+    public LspEditorFeatureDetector(
+        IVsService<SVsFeatureFlags, IVsFeatureFlags> vsFeatureFlagsService,
+        IVsService<SVsSettingsPersistenceManager, ISettingsManager> vsSettingsManagerService,
+        IUIContextService uiContextService,
+        JoinableTaskContext joinableTaskContext,
+        RazorActivityLog activityLog)
+    {
+        _uiContextService = uiContextService;
+        _jtf = joinableTaskContext.Factory;
+        _disposeTokenSource = new();
+
+        _lazyUseLegacyEditorTask = new(() =>
+             ComputeUseLegacyEditorAsync(vsFeatureFlagsService, vsSettingsManagerService, activityLog, _disposeTokenSource.Token),
+             _jtf);
+    }
+
+    public void Dispose()
+    {
+        if (_disposeTokenSource.IsCancellationRequested)
+        {
+            return;
+        }
+
+        _disposeTokenSource.Cancel();
+        _disposeTokenSource.Dispose();
+    }
+
+    private static async Task<bool> ComputeUseLegacyEditorAsync(
+        IVsService<SVsFeatureFlags, IVsFeatureFlags> vsFeatureFlagsService,
+        IVsService<SVsSettingsPersistenceManager, ISettingsManager> vsSettingsManagerService,
+        RazorActivityLog activityLog,
+        CancellationToken cancellationToken)
+    {
+        var vsFeatureFlags = await vsFeatureFlagsService.GetValueAsync(cancellationToken).ConfigureAwaitRunInline();
+
+        // IVsFeatureFlags is free-threaded but VSTHRD010 seems to be reported anyway.
+#pragma warning disable VSTHRD010 // Invoke single-threaded types on Main thread
+        if (vsFeatureFlags.IsFeatureEnabled(WellKnownFeatureFlagNames.UseLegacyRazorEditor, defaultValue: false))
+#pragma warning restore VSTHRD010 // Invoke single-threaded types on Main thread
+        {
+            activityLog.LogInfo($"Using Legacy Razor editor because the '{WellKnownFeatureFlagNames.UseLegacyRazorEditor}' feature flag is enabled.");
+            return true;
+        }
+
+        var settingsManager = await vsSettingsManagerService.GetValueAsync(cancellationToken).ConfigureAwaitRunInline();
+        var useLegacyEditorSetting = settingsManager.GetValueOrDefault<bool>(WellKnownSettingNames.UseLegacyASPNETCoreEditor);
+
+        if (useLegacyEditorSetting)
+        {
+            activityLog.LogInfo($"Using Legacy Razor editor because the '{WellKnownSettingNames.UseLegacyASPNETCoreEditor}' setting is set to true.");
+            return true;
+        }
+
+        activityLog.LogInfo($"Using LSP Razor editor.");
+        return false;
+    }
+
+    public bool IsLspEditorAvailable()
```

Comment:
I know you didn't come up with this name, but it's always annoyed me a little. Is it really checking if the editor is _available_? or is it actually checking if the editor is _enabled_?

Reply:
Totally agree!